### PR TITLE
ukify: Prefer using llvm-objcopy instead of objcopy

### PR DESCRIPTION
**Pull request by daandemeyer**
llvm-objcopy works on stubs built for foreign architectures whereas
objcopy doesn't so let's prefer using llvm-objcopy instead of objcopy.

llvm-objcopy automatically sets the virtual address and doesn't provide
an option to set it manually so we only add --change-section-vma when
using objcopy

The default section flags differ between llvm-objcopy and objcopy
so we add a default for the section flags so we make sure all sections
are read-only data unless specified otherwise.

### CODE REVIEW

**Comment by medhefgo** [comment]
Have you tested this?

```sh
$ llvm-objcopy systemd/build/src/boot/efi/linuxx64.efi.stub --add-section .osrel=/etc/os-release --add-section .linux=/boot/vmlinuz-linux --add-section .initrd=/boot/initramfs-linux.img /tmp/test.efi
$ objdump -h /tmp/test.efi 

/tmp/test.efi:     file format pei-x86-64

Sections:
Idx Name          Size      VMA               LMA               File off  Algn
  0 .text         00011850  0000000000003000  0000000000003000  00000400  2**4
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
  1 .reloc        0000000c  0000000000015000  0000000000015000  00011e00  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  2 .data         00005948  0000000000016000  0000000000016000  00012000  2**4
                  CONTENTS, ALLOC, LOAD, DATA
  3 .dynamic      00000110  000000000001c000  000000000001c000  00017a00  2**2
                  CONTENTS, ALLOC, LOAD, DATA
  4 .rela         00001008  000000000001d000  000000000001d000  00017c00  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  5 .dynsym       00000018  000000000001f000  000000000001f000  00018e00  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  6 .sbat         000000de  0000000000021000  0000000000021000  00019000  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  7 .sdmagic      00000035  00000000000210e0  00000000000210e0  00019200  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  8 .osrel        00000163  0000000000000000  0000000000000000  00019400  2**0
                  CONTENTS, ALLOC, LOAD, READONLY, DATA, NOREAD
  9 .linux        00a9e670  0000000000000000  0000000000000000  00019600  2**0
                  CONTENTS, ALLOC, LOAD, READONLY, DATA, NOREAD
 10 .initrd       00d07b7c  0000000000000000  0000000000000000  00ab7e00  2**0
                  CONTENTS, ALLOC, LOAD, READONLY, DATA, NOREAD
```

I remember it working when I tested this some months ago, might be arch dependent. What I do know is that it will pick the lowest available VMA. Which is a bad idea if you have a natively built efi binary (which is an end goal), as that creates images with a random(?) base VMA. So VMA @0 is always available, making the image probably unbootable (this might violate some PE rules and it very likely exhausts the memory space during image load).

**Comment by DaanDeMeyer** [comment]
You need to set the right section flags (data, readonly) to have it automatically calculate the right VMA

**Comment by medhefgo** [comment]
> You need to set the right section flags (data, readonly) to have it automatically calculate the right VMA.

Ah, that was it. llvm-objcopy is so weird. Also, it doesn't always work unless all new sections have their flags changed (I previously only checked by setting one of the sections).

I also quickly checked a natively built toy efi app and apparently, it sets the VMAs correctly and doesn't go below image base as long as the section flags are set.

So, I guess nvm my notes then…

**Comment by medhefgo** [comment]
On different note: it might be a nice if ukify verified the image for overlapping sections after generating it to catch any issues early in ci…

**Comment by DaanDeMeyer** [comment]
Added a commit to check for overlapping ranges

**Comment by DaanDeMeyer** [comment]
Also, while you might think this makes ukify able to stitch together aarch64 UKIs on x86, we need https://reviews.llvm.org/D140291 before that becomes possible.

File: src/ukify/ukify.py [hunk 1]
```diff
@@ -423,6 +424,21 @@ def join_initrds(initrds):
...
+def pe_validate(filename):
+    import pefile
+
+    pe = pefile.PE(filename)
+
+    sections = sorted(pe.sections, key=lambda s: (s.VirtualAddress, s.Misc_VirtualSize))
+
+    for i in range(len(sections) - 1):
+        l = sections[i]
+        r = sections[i + 1]
```
Comment (behrmann) [inline src/ukify/ukify.py]:
```suggestion
    for l, r in itertools.pairwise(sections):
```

File: src/ukify/ukify.py [hunk 2]
```diff
@@ -521,21 +521,22 @@ def make_uki(opts):
...
+             '--set-section-flags', f"{s.name}={','.join(s.flags)}")
             for s in uki.sections),
-        *itertools.chain.from_iterable(
-            ('--set-section-flags',  f"{s.name}={','.join(s.flags)}")
-            for s in uki.sections
-            if s.flags is not None),
         output,
     ]
+
+    if pathlib.Path(objcopy_tool).name == 'objcopy':
```
Comment (bluca) [inline src/ukify/ukify.py]: I'd recommend making this `!= llvm-objcopy` instead - reason being, usually to do cross-arch objcopy the binutils package might provide foreign target builds. For example on Debian you can install `binutils-aarch64-linux-gnu` which provides `aarch64-linux-gnu-objcopy` which would need `--change-section-vma`

**Review by bluca — APPROVED** [review]

File: src/ukify/ukify.py [hunk 3]
```diff
@@ -520,24 +533,27 @@ def make_uki(opts):
     else:
         output = opts.output
 
-    objcopy_tool = find_tool('objcopy', opts=opts)
+    objcopy_tool = find_tool('llvm-objcopy', 'objcopy', opts=opts)
```
Comment (medhefgo) [inline src/ukify/ukify.py:536]: I just found out that this only prefers llvm if `--tools` is passed.

**Pull request merged**